### PR TITLE
Speed up variant page loading

### LINKED ISSUE #1337: Scout too slow to work in
Scout has become so slow that it is no long possible to analyse our patients.  Some parts of Scout are working without any delay but when we click on a variant it can take up to one minute for the variant page to open.

### PR DESCRIPTION
**Pull request by northwestwitch**
This PR speeds up variant page loading (fix #1337)

**How to test**:
1. This branch is already on stage (you can also test before and after installing the branch)
1. Go to whichever variant and notice that page loads faster
1. Go to a variant that is causative also in another case: https://scout-stage.scilifelab.se/cust000/15026-miptest/e67092207773c559b7cbdbb6d4c521a9

**Expected outcome**:
1. Variant pages should load faster
1. Same variant causative in 2 cases should show correct link to causative variant in other case on top of variant page.

**Review:**
- [x] code approved by DN
- [x] tests executed by CR, DN

### CODE REVIEW

File: scout/server/blueprints/variants/controllers.py [hunk 1]
```diff
@@ -572,17 +571,14 @@ def variant(store, institute_obj, case_obj, variant_id=None, variant_obj=None, a
     for event in events:
         event['verb'] = VERBS_MAP[event['verb']]
 
-    other_causatives = []
+    other_causatives = set()
```
Comment (dnil) [inline scout/server/blueprints/variants/controllers.py]: 👍

File: scout/adapter/mongo/variant.py [hunk 2]
```diff
@@ -402,27 +402,50 @@ def check_causatives(self, case_obj=None, institute_obj=None, limit_genes=None):
...
-        institute_causatives = self.get_causatives(variant_obj['institute'])
-        for causative_id in institute_causatives:
-            other_variant = self.variant(causative_id)
-            if not other_variant:
+        variant_prefix = variant_obj['simple_id']
+        clinical_variant = ''.join([variant_prefix, '_clinical'])
+        research_variant = ''.join([variant_prefix, '_research'])
+
+        var_causative_events = self.event_collection.find({
+            'institute' : variant_obj['institute'],
```
Comment (dnil) [inline scout/adapter/mongo/variant.py]: Probably extend to collaborator level again?
The get_causatives version was 
```
       query = self.case_collection.aggregate([
                {'$match': {'collaborators': institute_id, 'causatives': {'$exists': True}}},
                {'$unwind': '$causatives'},
                {'$group': {'_id': '$causatives'}}
      causatives = [item['_id'] for item in query]
  ```

**Comment by northwestwitch** [comment]
> * previously flagged causatives seem to still be presented; seen on stage when looking eg at a validation case where the causatives are not currently flagged after rerun/reupload, but still present in the event log.

You don't remember which case this is?

**Comment by dnil** [comment]
There are some aspects of it for several of the validation cases. Try these:
https://scout-stage.scilifelab.se/cust002/F0007437-validation
https://scout-stage.scilifelab.se/cust002/F0009130/e4aec2a58c480520f215a49899a5221c

**Comment by northwestwitch** [comment]
> There are some aspects of it for several of the validation cases. Try these:
> https://scout-stage.scilifelab.se/cust002/F0007437-validation
> https://scout-stage.scilifelab.se/cust002/15030-validation
> https://scout-stage.scilifelab.se/cust002/F0009130/e4aec2a58c480520f215a49899a5221c

Ah ok now I understand, it's a fix in case page. I'll fix this as well!

File: scout/adapter/mongo/variant.py [hunk 3]
```diff
@@ -377,11 +377,8 @@ def check_causatives(self, case_obj=None, institute_obj=None, limit_genes=None):
             return []
 
         if case_obj:
-            # exclude variants that are marked causative in "case_obj"
-            case_causative_ids = set(case_obj.get('causatives', []))
-            institute_causative_variant_ids = list(
-                set(institute_causative_variant_ids).difference(case_causative_ids)
-            )
+            institute_causative_variant_ids = [id for id in institute_causative_variant_ids
+                if id not in case_obj.get('causatives',[])]
```
Comment (dnil) [inline scout/adapter/mongo/variant.py]: 👍

Comment (northwestwitch) [reply scout/adapter/mongo/variant.py]: I'm not done yet with this, Testing some stuff first!

**Comment by northwestwitch** [comment]
I might have fixed the problems of the 3 cases above

**Comment by dnil** [comment]
Yep, looks good to me!

File: tests/server/blueprints/cases/test_cases_views.py [hunk 4]
```diff
@@ -155,7 +155,7 @@ def test_causatives(app, user_obj, institute_obj, case_obj):
             institute=institute_obj,
             case=case_obj,
             user=user_obj,
-            link='causative_var_link',
+            link='causative_var_link/{}'.format(variant_obj['_id']),
```
Comment (dnil) [inline tests/server/blueprints/cases/test_cases_views.py]: Yep!

Comment (northwestwitch) [reply tests/server/blueprints/cases/test_cases_views.py]: Thanks! After the tests pass I'm going to merge!

Comment (dnil) [reply tests/server/blueprints/cases/test_cases_views.py]: 🤞

Comment (northwestwitch) [reply tests/server/blueprints/cases/test_cases_views.py]: YESSSS

**Pull request merged**